### ISSUE TITLE
Mongo pumps: Atlas connection issue (mongo TLS issue)

### ISSUE THREAD

**Issue by armujahid** [TykTechnologies/tyk-pump#540]
**Issue:**
Unable to connect with MongoDB 4.2.23 Enterprise Atlas cluster.

**Steps to reproduce:**
1) Create mongo 4.2 cluster on https://cloud.mongodb.com
2) You will get connection string like `mongodb+srv://<username>:<password>@something.xyz.mongodb.net/?retryWrites=true&w=majority` .Remove query params since pump throws error with `?retryWrites=true&w=majority`. Final connection string is `mongodb+srv://<username>:<password>@something.xyz.mongodb.net/tyk_analytics` that will be used in pump config
3) Clone https://github.com/TykTechnologies/tyk-pro-docker-demo
4) Optional: Update pump and dashboard version (Issue is also reproducible in latest versions)
5) Use `pump.mongo.env` to enable TLS (set `_MONGOUSESSL` and `_MONGOSSLINSECURESKIPVERIFY` for each pump. Although 2nd one is not required)
```
# Outdated variables in docs that doesn't work https://tyk.io/docs/tyk-pump/tyk-pump-configuration/tyk-pump-environment-variables/
# Docs should be updated
TYK_PMP_PUMPS_MONGO_META_MONGOUSESSL=true
TYK_PMP_PUMPS_MONGO_META_MONGOSSLINSECURESKIPVERIFY=true
TYK_PMP_PUMPS_MONGOAGGREGATE_META_MONGOUSESSL=true
TYK_PMP_PUMPS_MONGOAGGREGATE_META_MONGOSSLINSECURESKIPVERIFY=true
TYK_PMP_PUMPS_MONGOSELECTIVE_META_MONGOUSESSL=true
TYK_PMP_PUMPS_MONGOSELECTIVE_META_MONGOSSLINSECURESKIPVERIFY=true

TYK_PMP_PUMPS_MAIN_TYPE=mongo
TYK_PMP_PUMPS_MAIN_META_COLLECTIONNAME=tyk_analytics
TYK_PMP_PUMPS_MAIN_META_MONGOURL=mongodb+srv://<username>:<password>@something.xyz.mongodb.net/tyk_analytics
# added by me
TYK_PMP_PUMPS_MAIN_META_MONGOUSESSL=true
TYK_PMP_PUMPS_MAIN_META_MONGOSSLINSECURESKIPVERIFY=true

TYK_PMP_PUMPS_MAIN_META_MAXINSERTBATCHSIZEBYTES=80000
TYK_PMP_PUMPS_MAIN_META_MAXDOCUMENTSIZEBYTES=20112

TYK_PMP_PUMPS_MAINAGG_TYPE=mongo-pump-aggregate
TYK_PMP_PUMPS_MAINAGG_META_MONGOURL=mongodb+srv://<username>:<password>@something.xyz.mongodb.net/tyk_analytics
# added by me
TYK_PMP_PUMPS_MAINAGG_META_MONGOUSESSL=true
TYK_PMP_PUMPS_MAINAGG_META_MONGOSSLINSECURESKIPVERIFY=true

TYK_PMP_PUMPS_MAINAGG_META_USEMIXEDCOLLECTION=true
TYK_PMP_PUMPS_MAINAGG_META_TRACKALLPATHS=true

TYK_PMP_UPTIMEPUMPCONFIG_COLLECTIONNAME=tyk_uptime_analytics
TYK_PMP_UPTIMEPUMPCONFIG_MONGOURL=mongodb+srv://<username>:<password>@something.xyz.mongodb.net/tyk_analytics
# added by me
TYK_PMP_UPTIMEPUMPCONFIG_MONGOUSESSL=true
TYK_PMP_UPTIMEPUMPCONFIG_MONGOSSLINSECURESKIPVERIFY=true

TYK_PMP_UPTIMEPUMPCONFIG_MAXINSERTBATCHSIZEBYTES=500000
TYK_PMP_UPTIMEPUMPCONFIG_MAXDOCUMENTSIZEBYTES=200000

```
**Pump 1.7.0 output:**
```
time="Dec 06 13:37:12" level=info msg="## Tyk Analytics Pump, 1.7.0 ##" prefix=main
time="Dec 06 13:37:12" level=info msg="--> [REDIS] Creating cluster client"
time="Dec 06 13:37:12" level=info msg="Serving health check endpoint at http://localhost:8083/health ..." prefix=server
```
Below line is only logged if i use `mongodb://` instead of using `mongodb+srv://`
```
time="Dec 06 13:37:12" level=info msg=Init collection_name= prefix=mongo-pump url=
```

**Pump 1.6.0 output:**
```
time="Dec  5 06:55:18" level=info msg="## Tyk Analytics Pump, 1.6.0 ##" 
time="Dec  5 06:55:18" level=info msg="--> [REDIS] Creating cluster client" 
time="Dec  5 06:55:18" level=info msg="Serving health check endpoint at http://localhost:8083/health ..."
```
**Pump 1.5.1 output:**
```
time="Dec  5 07:18:02" level=info msg="## Tyk Analytics Pump, 1.5.1 ##" 
time="Dec  5 07:18:02" level=info msg="--> [REDIS] Creating cluster client" 
time="Dec  5 07:18:02" level=info msg="Serving health check endpoint at http://localhost:8083/health ..." 
time="Dec  5 07:18:02" level=info msg=Init collection_name= url=
```

**Additional Info:**
Mongo supported versions are MongoDB 3.x to 4.4.x as per https://tyk.io/docs/planning-for-production/database-settings/mongodb/#supported-versions

https://github.com/TykTechnologies/tyk-pump/pull/24 added TLS support for mongo but I think relevant environment variables are not working

Also facing similar issue with Tyk dashboard (even after setting `TYK_DB_MONGOUSESSL=true`)
```
time="Dec 05 07:21:47" level=info msg="Tyk Analytics Dashboard 4.2.3"
time="Dec 05 07:21:47" level=info msg="Copyright Tyk Technologies Ltd 2019"
time="Dec 05 07:21:47" level=info msg="https://www.tyk.io"
time="Dec 05 07:21:47" level=info msg="Using /opt/tyk-dashboard/tyk_analytics.conf for configuration"
time="Dec 05 07:21:47" level=info msg="Listening on port: 3000"
time="Dec 05 07:21:47" level=warning msg="Default admin_secret `12345` should be changed for production use."
time="Dec 05 07:21:47" level=info msg="Initializing Main storage with `mongo` database"
time="Dec 05 07:21:47" level=info msg="Connecting to MongoDB: [something.xyz.mongodb.net]"
time="2022-12-05T07:21:57Z" level=fatal msg="Could not connect to config store" error="no reachable servers"
```

P.S database is reachable since I can use `mongosh "<connection string>"` to connect to that exact DB

**Comment by armujahid** [comment]
As per [regex](https://regex101.com/r/8Uzwtw/1) mentioned at https://github.com/TykTechnologies/tyk-pump/pull/518/files atlas connection string started with `mongodb+srv://` is not supported :(

**Comment by oluwaseyeayinla** [comment]
You may be right about the srv. However, the section shared was introduced in the newly released pump 1.7.0. If srv is the issue then can you try [connecting without srv](https://stackoverflow.com/questions/56426020/how-to-get-standard-uri-connection-string-without-srv-on-mongodb-website-with)

**Comment by armujahid** [comment]
Connection string of this form `mongodb://<username>:<password>@some-cluster-01-shard-00-00.abcde.mongodb.net:27017,some-cluster-01-shard-00-01.abcde.mongodb.net:27017,some-cluster-01-shard-00-02.abcde.mongodb.net:27017/tyk_analytics?replicaSet=atlas-fghij-shard-0&authSource=admin` is working perfectly fine with pump If I set USESSL as
```
TYK_PMP_PUMPS_MAIN_META_MONGOUSESSL=true
TYK_PMP_PUMPS_MAINAGG_META_MONGOUSESSL=true
TYK_PMP_UPTIMEPUMPCONFIG_MONGOUSESSL=true
```

Issue is dashboard is still not working probably because `TYK_DB_MONGOUSESSL=true` as mentioned in docs is outdated (like outdated pump variables mentioned above).

Since this repo is only for pump that's why I am closing this issue. 
Dashboard issue is being discussed at https://community.tyk.io/t/pump-and-dashboard-unable-to-connect-with-mongodb-4-2-23-enterprise-atlas-cluster-tls-issue/6003
